q2_feature_classifier/_consensus_assignment.py

```diff
@@ -0,0 +1,252 @@
+# Copyright (c) 2016-2017, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import tempfile
+import subprocess
+import pandas as pd
+from os.path import isfile
+from collections import Counter
+
+
+def _get_default_unassignable_label():
+    return "Unassigned"
+
+
+def _consensus_assignments(
+        cmd, ref_taxa, min_consensus=0.51, output_no_hits=False,
+        exp_seq_ids=None,
+        unassignable_label=_get_default_unassignable_label()):
+    '''Run command line subprocess and find consensus taxonomy.'''
+    with tempfile.NamedTemporaryFile() as output:
+        cmd = cmd + [output.name]
+        _run_command(cmd)
+        obs_taxa = _import_blast_assignments(
+            output.name, ref_taxa, unassignable_label=unassignable_label)
+        consensus = _compute_consensus_annotations(
+            obs_taxa, min_consensus=min_consensus,
+            unassignable_label=unassignable_label)
+        # If output_no_hits=True, find/add query IDs missing from consensus
+        if output_no_hits is True:
+            consensus = _output_no_hits(
+                consensus, exp_seq_ids, unassignable_label)
+        # if there are no consensus results (no hits for any query), default
+        # to empty result file (must create empty dict or fail on pd error)
+        if not consensus:
+            consensus = {'': ('', '')}
+        result = pd.DataFrame.from_dict(consensus, 'index')
+        result.index.name = 'Feature ID'
+        result.columns = ['Taxon', 'Confidence']
+        return result
+
+
+def _output_no_hits(obs_taxa, exp_seq_ids,
+                    unassignable_label=_get_default_unassignable_label()):
+    '''If a query ID has no hits, report as unassigned.'''
+    # extract list of query sequence IDs
+    exp = _open_list_or_file(exp_seq_ids)
+    exp = [line.strip('>') for line in exp if line.startswith('>')]
+
+    # iterate over expected IDs, add as unassigned to obs_taxa if missing
+    for _id in exp:
+        if _id not in obs_taxa:
+            obs_taxa[_id] = (unassignable_label, 0.0)
+
+    return obs_taxa
+
+
+# Replace this function with QIIME2 API for wrapping commands/binaries,
+# pending https://github.com/qiime2/qiime2/issues/224
+def _run_command(cmd, verbose=True):
+    if verbose:
+        print("Running external command line application. This may print "
+              "messages to stdout and/or stderr.")
+        print("The command being run is below. This command cannot "
+              "be manually re-run as it will depend on temporary files that "
+              "no longer exist.")
+        print("\nCommand:", end=' ')
+        print(" ".join(cmd), end='\n\n')
+    subprocess.run(cmd, check=True)
+
+
+def _import_blast_assignments(
```

Comment:
I was also confused about this. @nbokulich, maybe a better name would be ``_assignments_from_bl7``, and then add a comment inside the top of the function noting that `bl7` refers to the BLAST+ output format 7?